feat: Backend for Temporal powered BatchExports

tomasfarias opened the pull request:
## Problem

NOTE: although this PR is closed, there is a refactored version that has been merged [here](https://github.com/PostHog/posthog/pull/15649)

We are building a new export platform for users who want to get their data out of PostHog and into their blob storage (like S3) and their data warehouses. In particular, we are looking to support batch exports that power ETL workflows, like the capabilities provided by other services (like [Fivetran](https://www.fivetran.com), or [Airbyte](https://airbyte.com/)), but without leaving PostHog.

<!-- Who are we building for, what are their needs, why is this important? -->

## Changes

This PR contains the backend for PostHog batch exports. These batch exports are built on top of [Temporal](https://www.temporal.io/): we leverage Temporal's scheduling and execution capabilities to run a new set of Batch Export workflows. Essentially, we have built a Batch Export abstraction layer on top of Temporal: Users need not worry about interfacing with Temporal (which can be quite complex), but rather only operate with predefined operations on the PostHog REST API.

The abstraction layer is composed of four models:
* `BatchExportDestination`: A model users will interact with to configure where to export data their data.
* `BatchExportSchedule`: A model mapping to a Temporal Schedule. It serves us to store Schedule configurations to pass to Temporal.
* `BatchExport`: The main abstraction users interact with. Each `BatchExport` represents the union of a `BatchExportSchedule` with a `BatchExportDestination`. In other words, a `BatchExport` is defined by a destination where data is meant to be exported, and a schedule to run said exports.
* `BatchExportRun`: Represents an instance of a `BatchExport`. A single `BatchExport` will have one `BatchExportRun` each time a workflow is triggered. These models are not meant to be created by users, but instead by Temporal, and they mostly serve for reporting on individual workflow executions.

The typical user would interact with the Batch Export abstraction layer by:
* Creating a `BatchExport` by making a POST request to `api/projects/:project_id/batch_export`. This POST request body includes a "schedule" and a "destination" key. The former specifies when the workflow will be ran, and it's validated against the `BatchExportSchedule`, whereas the latter specifies the configuration stored as a `BatchExportDestination`.

<!-- If there are frontend changes, please include screenshots. -->
<!-- If a reference design was involved, include a link to the relevant Figma frame! -->

## How did you test this code?

Unit tests for:
* New models in `posthog/models/tests/test_batch_export_model.py`.
* New API ViewSets in `posthog/api/tests/test_batch_exports.py`.
* New S3 Batch Export Workflow in `posthog/temporal/tests/test_s3_batch_export_workflow.py`.

Some tests use the `object-storage` MinIO bucket deployed as part of the development stack to export data as if it were an S3 Bucket (as the API is compatible). Then we compare the data exported to ensure it matches the events expected.

## Open questions/TODOs

* Should we limit columns exported? We are currently exporting all (`*`).
* Retrying individual `BatchExportRun`s is still WIP. Temporal doesn't expose a `retry` method directly, only via a separate `Replayer` Worker. Spawning a separate worker seems too complex for a task that I expect to be common, so we'll likely need to implement a signal-based approach.

<!-- Briefly describe the steps you took. -->
<!-- Include automated tests if possible, otherwise describe the manual testing routine. -->

tomasfarias committed:
feat: Backend for Temporal powered BatchExports

tomasfarias commented:
Rebased to get latest master branch changes.

hazzadous commented on posthog/temporal/workflows/s3_batch_export.py:
```diff
@@ -0,0 +1,342 @@
+    else:
+        key = key_template.format(**template_vars)
+
+    if TEST or DEBUG:
```
Can we use `OBJECT_STORAGE_ENDPOINT` here instead?

tomasfarias replied:
Not quite: `settings.OBJECT_STORAGE_ENDPOINT` is set to `http://localhost:19000` which is accessible via the host. However, the ClickHouse container does not share the network from the host, so it uses the service name. I made a comment explaining this right below.

hazzadous replied:
ah yup yup 💩

hazzadous commented on posthog/models/test/test_batch_export_model.py:
```diff
@@ -0,0 +1,362 @@
+    BatchExportRun,
+    BatchExportSchedule,
+)
+from posthog.settings import (
```
I think we should be using `from django.conf import settings` here right? As the canonical Django way. Or does that not really matter here?

neilkakkar replied:
the only thing it matters for is testing: `override_settings` works only with `from django.conf import settings`, while in the other case, you'll have to patch the setting itself. Makes it convenient to use django.conf

tomasfarias replied:
I meant that it makes no difference in this PR, but yeah, I do see the value of `override_settings` and just overall being consistent with everything else 👍 I've made the switch to use `django.conf.settings` and used `override_settings` instead of `mock` patching in the test file.

hazzadous commented on posthog/api/batch_exports.py:
```diff
@@ -0,0 +1,224 @@
+
+        user_id = request.user.distinct_id
+        team_id = request.user.current_team.id
+        note = f"Unpause requested by user {user_id} from team {team_id}"
```
We have an "activity log" used for other data type changes, thinking maybe it's relevant here as well. Not important for this PR though.

hazzadous reviewed (changes requested):
Looks good

Like you mentioned, some API tests for 401/403 are needed. And we need to wait for the migration.

hazzadous commented on posthog/temporal/workflows/s3_batch_export.py:
```diff
@@ -0,0 +1,342 @@
+
+INSERT_INTO_S3_QUERY_TEMPLATE = Template(
+    """
+    INSERT INTO FUNCTION s3({path}, $auth {file_format})
```
Also, to register my objections, I would rather we handle the S3 upload ourselves than rely on ClickHouse here. The removal of a hop is nice and all, but debugging things in ClickHouse has been tricky, and I'd take slow and steady here over a marginal reduction in network traffic.

It also makes the management of ClickHouse a little annoying as we're now coupling the app functionality with ClickHouse config for not enough gain imo.

hazzadous replied:
(But we can change this outside of this PR)

tomasfarias replied:
A fair objection 👍

If this becomes becomes an issue in the future more than happy to do the extra work. Very little would need to be deleted as we still need to select the data, configure the destination, schedule the workflow... So I think we can build on top of this as we discover things

tomasfarias committed:
refactor: Use django settings instead of importing from module

This required a small refactor in tests as we were mocking the
settings module.

Also re-add add-search-attributes service to docker-compose as it was
accidentally removed before.

tomasfarias commented:
Rebased on `master` to clear some conflicts with requirements files.

hazzadous commented:
there's a merge conflict 😱

hazzadous commented:
Also a note that we need to wait for the migration to complete, which if I recall correctly is today?!

tomasfarias committed:
refactor: Encrypt BatchExportDestination config in the database (#15577)

tomasfarias committed:
fix: Squash migrations 0315 and 0314

tomasfarias committed:
Merge branch 'master' into feat/batch-exports-backend-by-temporal

tomasfarias committed:
fix: Do not repeat activity

hazzadous commented:
This has been refactored and merged in https://github.com/PostHog/posthog/pull/15649
